lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -649,32 +608,119 @@ void TypeChecker::validateGenericTypeSignature(GenericTypeDecl *typeDecl) {
+      return subscr->getGenericSignature();
+    }
+    llvm_unreachable("Wat?");
   }
 
-  assert(!typeDecl->getGenericEnvironment());
+  bool allowConcreteGenericParams = false;
+  SmallVector<TypeLoc, 2> inferenceSources;
+  SmallVector<Requirement, 2> sameTypeReqs;
+  if (auto VD = dyn_cast_or_null<ValueDecl>(GC->getAsDecl())) {
+    auto func = dyn_cast<AbstractFunctionDecl>(VD);
+    auto subscr = dyn_cast<SubscriptDecl>(VD);
+
+    // For functions and subscripts, resolve the parameter and result types and
+    // note them as inference sources.
+    if (subscr || func) {
+      // Gather requirements from the parameter list.
+      auto resolution = TypeResolution::forStructural(GC);
+
+      TypeResolutionOptions options =
+          (func ? TypeResolverContext::AbstractFunctionDecl
+                : TypeResolverContext::SubscriptDecl);
+
+      auto params = func ? func->getParameters() : subscr->getIndices();
+      for (auto param : *params) {
+        auto *typeRepr = param->getTypeLoc().getTypeRepr();
+        if (typeRepr == nullptr)
+          continue;
 
-  // We don't go down this path for protocols; instead, the generic signature
-  // is simple enough that GenericContext::getGenericSignature() can build it
-  // directly.
-  assert(!isa<ProtocolDecl>(typeDecl));
+        auto paramOptions = options;
+        paramOptions.setContext(param->isVariadic()
+                                    ? TypeResolverContext::VariadicFunctionInput
+                                    : TypeResolverContext::FunctionInput);
+        paramOptions |= TypeResolutionFlags::Direct;
 
-  auto *gp = typeDecl->getGenericParams();
-  auto *dc = typeDecl->getDeclContext();
+        auto type = resolution.resolveType(typeRepr, paramOptions);
 
-  if (!gp) {
-    typeDecl->setGenericSignature(dc->getGenericSignatureOfContext());
-    return;
-  }
+        if (auto *specifier = dyn_cast<SpecifierTypeRepr>(typeRepr))
+          typeRepr = specifier->getBase();
+
+        inferenceSources.emplace_back(typeRepr, type);
+      }
+
+      // Gather requirements from the result type.
+      auto *resultTypeRepr = [&subscr, &func]() -> TypeRepr * {
+        if (subscr) {
+          return subscr->getElementTypeLoc().getTypeRepr();
+        } else if (auto *FD = dyn_cast<FuncDecl>(func)) {
+          return FD->getBodyResultTypeLoc().getTypeRepr();
+        } else {
+          return nullptr;
+        }
+      }();
+      if (resultTypeRepr && !isa<OpaqueReturnTypeRepr>(resultTypeRepr)) {
+        auto resultType = resolution.resolveType(
+            resultTypeRepr, TypeResolverContext::FunctionResult);
+
+        inferenceSources.emplace_back(resultTypeRepr, resultType);
+      }
+    }
+  } else if (auto *ext = dyn_cast<ExtensionDecl>(GC)) {
+    // Form the interface type of the extension so we can use it as an inference
+    // source.
+    //
+    // FIXME: Push this into the "get interface type" request.
```

Comment:
Actually I'd like to redo how the inference works here so that formExtensionInterfaceType() is not needed at all.